### PR TITLE
segment: add for provisioning API

### PR DESCRIPTION
**Pull request by sumnerevans**

### CODE REVIEW

File: config/upgrade.go [hunk 1]
```diff
@@ -54,6 +54,7 @@ func (helper *UpgradeHelper) doUpgrade() {
 	} else {
 		helper.Copy(Str, "appservice", "provisioning", "shared_secret")
 	}
+	helper.Copy(Str, "appservice", "provisioning", "segment_key")
```
Comment (tulir) [inline config/upgrade.go:57]:
```suggestion
	helper.Copy(Str|Null, "appservice", "provisioning", "segment_key")
```

File: segment.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+
+	log "maunium.net/go/maulogger/v2"
+	"maunium.net/go/mautrix/id"
+)
+
+var SEGMENT_URL string = "https://api.segment.io/v1/track"
```
Comment (tulir) [inline segment.go:28]: Go doesn't use uppercase for anything, and this should probably be a constant.

```suggestion
const SegmentURL = "https://api.segment.io/v1/track"
```

File: segment.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	if len(properties) > 1 {
+		segment.log.Fatalf("Track should be called with at most one property map")
+	}
+
+	go (func() error {
+		props := map[string]interface{}{}
+		if len(properties) > 0 {
+			props = properties[0]
+		}
```
Comment (tulir) [inline segment.go:56]: Long inline functions aren't very nice, this should probably be moved to a separate private (lowercase) function

**Review by tulir — COMMENTED** [review]
It's technically possible to also track exactly when the QR code is scanned rather than only when the login succeeds or fails, but that might be too much effort since we haven't had any problems with going from scanned to success.

File: segment.go (hunk 3 above)
Comment (sumnerevans) [reply segment.go:80]: I kept a small part of it (for wrangling the properties and handling errors), but extracted out the meat of it

**Pull request merged**